ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/EpochProcessorUtil.java

```diff
@@ -444,17 +444,22 @@ public static void attestionInclusion(BeaconState state, UnsignedLong previous_t
       throws EpochProcessingException {
     try {
       List<Integer> previous_indices = AttestationUtil.get_previous_epoch_attester_indices(state);
-      for (int index : previous_indices) {
-        UnsignedLong inclusion_slot = AttestationUtil.inclusion_slot(state, index);
-        int proposer_index = BeaconStateUtil.get_beacon_proposer_index(state, inclusion_slot);
-        List<UnsignedLong> balances = state.getValidator_balances();
-        UnsignedLong balance = balances.get(proposer_index);
-        UnsignedLong reward =
-            base_reward(state, index, previous_total_balance)
-                .dividedBy(UnsignedLong.valueOf(Constants.ATTESTATION_INCLUSION_REWARD_QUOTIENT));
-        balance = balance.plus(reward);
-        balances.set(proposer_index, balance);
-      }
+      previous_indices
+          .parallelStream()
+          .forEach(
+              index -> {
+                UnsignedLong inclusion_slot = AttestationUtil.inclusion_slot(state, index);
+                int proposer_index =
+                    BeaconStateUtil.get_beacon_proposer_index(state, inclusion_slot);
+                List<UnsignedLong> balances = state.getValidator_balances();
+                UnsignedLong balance = balances.get(proposer_index);
```

Comment:
Does Java Stream API handle race conditions that might arise when parallel processing the balance for the same proposer_index in two different streams? or should we be worried about that?

Reply:
I need to look through these again. In general, there is no magic bullet. You should always make sure that the state of one element does not affect another element.

Reply:
👍